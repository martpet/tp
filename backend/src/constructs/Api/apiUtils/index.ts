export * from './cookieName';
export * from './errorResponse';
export * from './handlerWrapper';
export * from './itResolvesWithApiError';
