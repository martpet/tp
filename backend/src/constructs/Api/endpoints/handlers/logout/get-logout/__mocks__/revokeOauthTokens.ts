export const revokeOauthTokens = vi.fn().mockResolvedValue(undefined);
