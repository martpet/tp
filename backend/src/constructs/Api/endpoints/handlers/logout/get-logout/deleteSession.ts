import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DeleteCommand, DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';

import { sessionsTableOptions } from '~/consts';
import { SessionsTableItem } from '~/types';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

export const deleteSession = async (sessionId: string) => {
  const deleteCommand = new DeleteCommand({
    TableName: sessionsTableOptions.tableName,
    Key: { [sessionsTableOptions.partitionKey.name]: sessionId },
    ReturnValues: 'ALL_OLD',
  });

  const { Attributes } = await ddbDocClient.send(deleteCommand);
  const sessionsItem = Attributes as SessionsTableItem | undefined;

  if (!sessionsItem) {
    throw new Error(`Could not find "Sessions" item with id "${sessionId}"`);
  }

  return sessionsItem.refreshToken;
};
