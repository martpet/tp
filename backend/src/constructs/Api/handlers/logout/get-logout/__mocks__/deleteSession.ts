export const deleteSession = vi.fn().mockResolvedValue('dummyRefreshToken');
