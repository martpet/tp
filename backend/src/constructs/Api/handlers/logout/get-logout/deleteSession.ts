import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DeleteCommand, DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';

import { sessionsTableOptions } from '~/consts';
import { SessionsTableItem } from '~/types';

const ddbClient = new DynamoDBClient({});
const ddbDocumentClient = DynamoDBDocumentClient.from(ddbClient);

export const deleteSession = async (sessionId: string) => {
  const deleteCommand = new DeleteCommand({
    TableName: sessionsTableOptions.tableName,
    Key: { [sessionsTableOptions.partitionKey.name]: sessionId },
    ReturnValues: 'ALL_OLD',
  });

  const { Attributes } = await ddbDocumentClient.send(deleteCommand);
  const item = Attributes as SessionsTableItem | undefined;

  return item!.refreshToken;
};
