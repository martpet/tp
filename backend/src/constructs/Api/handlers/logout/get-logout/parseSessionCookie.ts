import cookie from 'cookie';

import { cookieName } from '../../../apiUtils';

export const parseSessionCookie = (cookies?: string[]) => {
  try {
    const parsedCookies = cookie.parse(cookies!.join(';'));
    return parsedCookies[cookieName('session')];
  } catch (e) {
    console.error(e);
    return undefined;
  }
};
