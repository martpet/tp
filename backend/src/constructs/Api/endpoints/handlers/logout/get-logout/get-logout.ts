import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import cookie from 'cookie';
import { StatusCodes } from 'http-status-codes';

import { ApiRouteEnvVars, ApiRouteHeaders } from '~/constructs/Api/types';
import { cookieName, errorResponse, parseEventCookies } from '~/constructs/Api/utils';
import { authPaths, localhostUrl } from '~/consts';

import { deleteSession } from './deleteSession';
import { revokeOauthTokens } from './revokeOauthTokens';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const { envName } = globalLambdaProps;
  const { authDomain, clientId, logoutCallbackUrl, logoutCallbackLocalhostUrl } =
    process.env as ApiRouteEnvVars<'/logout'>;
  const { referer } = event.headers as ApiRouteHeaders<'/logout'>;
  const { sessionId } = parseEventCookies<'/logout'>(event);
  const isFromLocalhost = envName === 'personal' && referer?.startsWith(localhostUrl);
  const cognitoLogoutUrl = new URL(`https://${authDomain}${authPaths.logout}`);
  const blankSessionCookie = cookie.serialize(cookieName('sessionId'), '', {
    expires: new Date(),
  });

  if (!authDomain || !clientId || !logoutCallbackUrl || !logoutCallbackLocalhostUrl) {
    return errorResponse('WKSoIIT1Ds');
  }

  if (envName === 'personal' && !referer) {
    return errorResponse('GkjK-mpVvL');
  }

  if (sessionId) {
    let refreshToken: string;

    try {
      refreshToken = await deleteSession(sessionId);
    } catch (error) {
      return errorResponse('If_8Sr_Y7L', { error });
    }

    try {
      await revokeOauthTokens({ authDomain, clientId, refreshToken });
    } catch (error) {
      return errorResponse('-8-K2WmZuo', { error });
    }
  }

  const cognitoLogoutUrlParams = new URLSearchParams({
    client_id: clientId,
    logout_uri: isFromLocalhost ? logoutCallbackLocalhostUrl : logoutCallbackUrl,
  });

  cognitoLogoutUrl.search = cognitoLogoutUrlParams.toString();

  return {
    statusCode: StatusCodes.MOVED_TEMPORARILY,
    headers: { location: cognitoLogoutUrl.href },
    cookies: [blankSessionCookie],
  };
};
