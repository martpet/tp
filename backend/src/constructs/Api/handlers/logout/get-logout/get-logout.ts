import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import cookie from 'cookie';
import { StatusCodes } from 'http-status-codes';

import { authPaths, localhostUrl } from '~/consts';
import { EnvName } from '~/types';

import { ProcessEnv, RequestHeaders } from '../../../apiTypes';
import { cookieName, errorResponse } from '../../../apiUtils';
import { deleteSession } from './deleteSession';
import { parseSessionCookie } from './parseSessionCookie';
import { revokeOauthTokens } from './revokeOauthTokens';

export const handler: APIGatewayProxyHandlerV2 = async ({ cookies, headers }) => {
  const { authDomain, clientId, logoutCallbackUrl, logoutCallbackLocalhostUrl, envName } =
    process.env as ProcessEnv<'/logout'>;

  const { referer } = headers as RequestHeaders<'/logout'>;

  if (
    !authDomain ||
    !clientId ||
    !logoutCallbackUrl ||
    !logoutCallbackLocalhostUrl ||
    !envName
  ) {
    return errorResponse('WKSoIIT1Ds');
  }

  const isPersonalEnv = (envName as EnvName) === 'personal';
  const isFromLocalhost = isPersonalEnv && referer?.startsWith(localhostUrl);

  if (isPersonalEnv && !referer) {
    return errorResponse('GkjK-mpVvL');
  }

  const sessionId = parseSessionCookie(cookies);

  if (sessionId) {
    let refreshToken: string;

    try {
      refreshToken = await deleteSession(sessionId);
    } catch (error) {
      return errorResponse('If_8Sr_Y7L', { error });
    }

    try {
      await revokeOauthTokens({ authDomain, clientId, refreshToken });
    } catch (error) {
      return errorResponse('-8-K2WmZuo', { error });
    }
  }

  const cognitoLogoutUrl = new URL(`https://${authDomain}${authPaths.logout}`);

  const cognitoLogoutUrlParams = new URLSearchParams({
    client_id: clientId,
    logout_uri: isFromLocalhost ? logoutCallbackLocalhostUrl : logoutCallbackUrl,
  });

  cognitoLogoutUrl.search = cognitoLogoutUrlParams.toString();

  const blankSessionCookie = cookie.serialize(cookieName('session'), '', {
    expires: new Date(),
  });

  return {
    statusCode: StatusCodes.MOVED_TEMPORARILY,
    headers: { location: cognitoLogoutUrl.href },
    cookies: [blankSessionCookie],
  };
};
