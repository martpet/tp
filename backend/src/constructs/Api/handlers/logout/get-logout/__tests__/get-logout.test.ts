import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import cookie from 'cookie';

import { itResolvesWithApiError } from '../../../../apiUtils';
import { deleteSession } from '../deleteSession';
import { handler } from '../get-logout';
import { parseSessionCookie } from '../parseSessionCookie';
import { revokeOauthTokens } from '../revokeOauthTokens';

vi.mock('cookie');
vi.mock('../deleteSession');
vi.mock('../parseSessionCookie');
vi.mock('../revokeOauthTokens');
vi.mock('../../../../apiUtils/errorResponse');

const args = [
  {
    cookies: 'dummyCookies',
    headers: {
      referer: 'dummyRefererHeader',
    },
  },
] as unknown as Parameters<APIGatewayProxyHandlerV2>;

beforeEach(() => {
  process.env.authDomain = 'dummyAuthDomain';
  process.env.clientId = 'dummyClientId';
  process.env.logoutCallbackUrl = 'dummyLogoutCallbackUrl';
  process.env.logoutCallbackLocalhostUrl = 'dummyLogoutCallbackLocalhostUrl';
  process.env.envName = 'dummyEnvName';
});

describe('"get-logout" handler', () => {
  it('calls "cookie.serialize" with correct args', async () => {
    await handler(...args);
    expect(vi.mocked(cookie.serialize).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(handler(...args)).resolves.toMatchSnapshot();
  });

  describe('when "parseSessionCookie" returns a value', () => {
    beforeEach(() => {
      vi.mocked(parseSessionCookie).mockReturnValueOnce('dummySessionId');
    });

    it('calls "deleteSession" with correct args', async () => {
      await handler(...args);
      expect(vi.mocked(deleteSession).mock.calls).toMatchSnapshot();
    });

    it('calls "revokeOauthTokens" with correct args', async () => {
      await handler(...args);
      expect(vi.mocked(revokeOauthTokens).mock.calls).toMatchSnapshot();
    });

    describe('when "deleteSession" rejects', () => {
      beforeEach(() => {
        vi.mocked(deleteSession).mockRejectedValueOnce('dummyDeleteSessionRejectedValue');
      });
      itResolvesWithApiError(handler, args);
    });

    describe('when "revokeOauthTokens" rejects', () => {
      beforeEach(() => {
        vi.mocked(revokeOauthTokens).mockRejectedValueOnce(
          'dummyRevokeOauthTokensRejectedValue'
        );
      });
      itResolvesWithApiError(handler, args);
    });
  });

  describe('when "envName" is "personal"', () => {
    beforeEach(() => {
      process.env.envName = 'personal';
    });

    describe('when "referer" header is missing', () => {
      const argsClone = structuredClone(args);
      delete argsClone[0].headers.referer;

      itResolvesWithApiError(handler, argsClone);
    });

    describe('when "referer" is from localhost', () => {
      const argsClone = structuredClone(args);
      argsClone[0].headers.referer = 'http://localhost:3000/dummyPath';

      it('resolves with a correct value', () => {
        return expect(handler(...argsClone)).resolves.toMatchSnapshot();
      });
    });
  });

  describe.each([
    'authDomain',
    'clientId',
    'logoutCallbackUrl',
    'logoutCallbackLocalhostUrl',
    'envName',
  ])('when "%s" env var is missing', (key) => {
    beforeEach(() => {
      delete process.env[key];
    });
    itResolvesWithApiError(handler, args);
  });
});
