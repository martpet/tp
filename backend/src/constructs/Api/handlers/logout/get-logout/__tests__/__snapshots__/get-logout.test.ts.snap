// Vitest Snapshot v1

exports[`"get-logout" handler > calls "cookie.serialize" with correct args 1`] = `
[
  [
    "session",
    "",
    {
      "expires": 2022-01-01T00:00:00.000Z,
    },
  ],
]
`;

exports[`"get-logout" handler > resolves with a correct value 1`] = `
{
  "cookies": [
    "dummyCookie",
  ],
  "headers": {
    "location": "https://dummyauthdomain/logout?client_id=dummyClientId&logout_uri=dummyLogoutCallbackUrl",
  },
  "statusCode": 302,
}
`;

exports[`"get-logout" handler > when "authDomain" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "WKSoIIT1Ds",
  ],
]
`;

exports[`"get-logout" handler > when "authDomain" env var is missing > resolves with a correct value 1`] = `"dummyerrorResponse"`;

exports[`"get-logout" handler > when "clientId" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "WKSoIIT1Ds",
  ],
]
`;

exports[`"get-logout" handler > when "clientId" env var is missing > resolves with a correct value 1`] = `"dummyerrorResponse"`;

exports[`"get-logout" handler > when "envName" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "WKSoIIT1Ds",
  ],
]
`;

exports[`"get-logout" handler > when "envName" env var is missing > resolves with a correct value 1`] = `"dummyerrorResponse"`;

exports[`"get-logout" handler > when "envName" is "personal" > when "referer" header is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "GkjK-mpVvL",
  ],
]
`;

exports[`"get-logout" handler > when "envName" is "personal" > when "referer" header is missing > resolves with a correct value 1`] = `"dummyerrorResponse"`;

exports[`"get-logout" handler > when "envName" is "personal" > when "referer" is from localhost > resolves with a correct value 1`] = `
{
  "cookies": [
    "dummyCookie",
  ],
  "headers": {
    "location": "https://dummyauthdomain/logout?client_id=dummyClientId&logout_uri=dummyLogoutCallbackLocalhostUrl",
  },
  "statusCode": 302,
}
`;

exports[`"get-logout" handler > when "logoutCallbackLocalhostUrl" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "WKSoIIT1Ds",
  ],
]
`;

exports[`"get-logout" handler > when "logoutCallbackLocalhostUrl" env var is missing > resolves with a correct value 1`] = `"dummyerrorResponse"`;

exports[`"get-logout" handler > when "logoutCallbackUrl" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "WKSoIIT1Ds",
  ],
]
`;

exports[`"get-logout" handler > when "logoutCallbackUrl" env var is missing > resolves with a correct value 1`] = `"dummyerrorResponse"`;

exports[`"get-logout" handler > when "parseSessionCookie" returns a value > calls "deleteSession" with correct args 1`] = `
[
  [
    "dummySessionId",
  ],
]
`;

exports[`"get-logout" handler > when "parseSessionCookie" returns a value > calls "revokeOauthTokens" with correct args 1`] = `
[
  [
    {
      "authDomain": "dummyAuthDomain",
      "clientId": "dummyClientId",
      "refreshToken": "dummyRefreshToken",
    },
  ],
]
`;

exports[`"get-logout" handler > when "parseSessionCookie" returns a value > when "deleteSession" rejects > calls "errorResponse" with correct args 1`] = `
[
  [
    "If_8Sr_Y7L",
    {
      "error": "dummyDeleteSessionRejectedValue",
    },
  ],
]
`;

exports[`"get-logout" handler > when "parseSessionCookie" returns a value > when "deleteSession" rejects > resolves with a correct value 1`] = `"dummyerrorResponse"`;

exports[`"get-logout" handler > when "parseSessionCookie" returns a value > when "revokeOauthTokens" rejects > calls "errorResponse" with correct args 1`] = `
[
  [
    "-8-K2WmZuo",
    {
      "error": "dummyRevokeOauthTokensRejectedValue",
    },
  ],
]
`;

exports[`"get-logout" handler > when "parseSessionCookie" returns a value > when "revokeOauthTokens" rejects > resolves with a correct value 1`] = `"dummyerrorResponse"`;
